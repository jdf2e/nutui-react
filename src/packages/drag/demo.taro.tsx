import React from 'react'
import Taro from '@tarojs/taro'
import { Drag, Button } from '@/packages/nutui.react.taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import Header from '@/sites/components/header'

interface T {
  basic: string
  dragBasic: string
  direction: string
  directionX: string
  directionY: string
  attract: string
  attractText: string
  limitBoundaries: string
}
const DragDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      basic: '基础用法',
      dragBasic: '触摸移动',
      direction: '限制拖拽方向',
      directionX: '只能X轴拖动',
      directionY: '只能Y轴拖动',
      attract: '自动吸边',
      attractText: '按钮',
      limitBoundaries: '限制拖拽边界',
    },
    'zh-TW': {
      basic: '基础用法',
      dragBasic: '觸摸移動',
      direction: '限製拖拽方向',
      directionX: '只能X軸拖動',
      directionY: '只能Y軸拖動',
      attract: '自動吸邊',
      attractText: '按鈕',
      limitBoundaries: '限製拖拽邊界',
    },
    'en-US': {
      basic: 'Basic Usage',
      dragBasic: 'Button',
      direction: 'Limit Direction',
      directionX: 'X axis',
      directionY: 'Y axis',
      attract: 'Attract',
      attractText: 'Button',
      limitBoundaries: 'Limit Boundaries',
    },
  })
  const { screenWidth, windowHeight } = Taro.getSystemInfoSync()

  const isTaroWeb = Taro.getEnv() === 'WEB'

  const right = () => {
    return screenWidth - 300 - 9
  }
  const bottom = () => {
    return windowHeight - 501 - (isTaroWeb ? 57 : 0)
  }
  return (
    <>
      <Header />
      <div className={`demo ${isTaroWeb ? 'web' : ''}`}>
        <h2>{translated.basic}</h2>
        <Drag style={{ left: '8px' }}>
          <Button type="primary">{translated.dragBasic}</Button>
        </Drag>
        <h2 style={{ top: '30px', position: 'relative' }}>
          {translated.direction}
        </h2>
        <Drag
          direction="x"
          style={{
            top: isTaroWeb ? '197px' : '140px',
            left: '8px',
          }}
        >
          <Button type="primary"> {translated.directionX}</Button>
        </Drag>
        <Drag
          direction="y"
          style={{
            top: isTaroWeb ? '197px' : '140px',
            right: '50px',
          }}
        >
          <Button type="primary"> {translated.directionY}</Button>
        </Drag>
        <h2 style={{ top: '60px', position: 'relative' }}>
          {translated.attract}
        </h2>
        <Drag
          direction="x"
          attract
          style={{
            top: isTaroWeb ? '277px' : '220px',
            left: '8px',
          }}
        >
          <Button type="primary">{translated.attractText}</Button>
        </Drag>
        <h2 style={{ top: '90px', position: 'relative' }}>
          {translated.limitBoundaries}
        </h2>
        <div
          className="drag-boundary"
          style={{
            position: 'absolute',
            top: isTaroWeb ? '357px' : '300px',
            left: '8px',
            width: '300px',
            height: '200px',
            border: '1px solid var(--nutui-color-primary)',
          }}
        />
        <Drag
          boundary={{
            top: isTaroWeb ? 358 : 301,
            left: 9,
            bottom: bottom(),
            right: right(),
          }}
          style={{
            top: isTaroWeb ? '417px' : '360px',
            left: '50px',
          }}
        >
          <Button type="primary">{translated.limitBoundaries}</Button>
        </Drag>
      </div>
    </>
  )
}

export default DragDemo
