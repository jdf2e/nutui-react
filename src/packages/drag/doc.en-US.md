# Drag 

### Intro

Implement draggable arbitrary elements.

### Install

```javascript
import { Drag } from '@nutui/nutui-react'
```

## Demo

### Basic Usage

:::demo

```tsx
import  React from "react";
import { Drag } from '@nutui/nutui-react';

const App = () => {
  const btnStyle = {
    borderRadius: '25px',
    padding: '0 18px',
    fontSize: '14px',
    color: '#fff',
    display: 'inline-block',
    lineHeight: '36px',
    background: 'linear-gradient(135deg,#fa2c19 0,#fa6419 100%)',
  }
  return (
    <Drag>
        <div className="touch-dom" style={btnStyle}>Button</div>
    </Drag>
 );
};
export default App;
```

:::

##  Limit Direction

:::demo

```tsx
import  React from "react";
import { Drag } from '@nutui/nutui-react';

const App = () => {
  const btnStyle = {
    borderRadius: '25px',
    padding: '0 18px',
    fontSize: '14px',
    color: '#fff',
    display: 'inline-block',
    lineHeight: '36px',
    background: 'linear-gradient(135deg,#fa2c19 0,#fa6419 100%)',
  }
  return (
    <>
    <Drag direction="x" style={{ top: '200px', left: '8px' }}>
        <span style={btnStyle}>X axis</span>
      </Drag>
      <Drag direction="y" style={{ top: '200px', right: '50px' }}>
        <span style={btnStyle}>Y axis</span>
      </Drag>
    </>
 );
};
export default App;
```

:::

## Attract

:::demo

```tsx
import  React from "react";
import { Drag } from '@nutui/nutui-react';

const App = () => {
  const btnStyle = {
    borderRadius: '25px',
    padding: '0 18px',
    fontSize: '14px',
    color: '#fff',
    display: 'inline-block',
    lineHeight: '36px',
    background: 'linear-gradient(135deg,#fa2c19 0,#fa6419 100%)',
  }
  return (
    <Drag direction="x" attract>
      <div className="touch-dom" style={btnStyle}>Button</div>
    </Drag>
 );
};
export default App;
```

:::

## Limit Boundaries
:::demo

```tsx
import  React from "react";
import { Drag } from '@nutui/nutui-react';

const App = () => {
  const btnStyle = {
    borderRadius: '25px',
    padding: '0 18px',
    fontSize: '14px',
    color: '#fff',
    display: 'inline-block',
    lineHeight: '36px',
    background: 'linear-gradient(135deg,#fa2c19 0,#fa6419 100%)',
  }
  
    const right = () => {
    return document.documentElement.clientWidth - 300 - 9
    }
    const bottom = () => {
        return document.documentElement.clientHeight - 202
    }
    return (
    <>
    <div
        className="drag-boundary"
        style={{
          position: 'absolute',
          top: '0px',
          left: '8px',
          width: '300px',
          height: '200px',
          border: '1px solid red',
        }}
       />
      <Drag
        boundary={{ top: 1, left: 9, bottom: bottom(), right: right() }}
        style={{ top: '40px', left: '50px' }}
      >
        <span style={btnStyle}>Limit Boundaries</span>
      </Drag>
    </>
   );
};
export default App;
```

:::

## API

### Props


| Attribute            | Description               | Type   | Default  |
| :-------- | :------------------------------------------------ | :------------- | :---------------------------------- |
| attract   | Whether to enable automatic edge suction  | Boolean        | false                                |
| direction | The drag direction limit of the dragged element **x**/**y**/**all**| String   | 'all'         |
| boundary  | The drag boundary of the dragged element   | Object         | {top: 0,left: 0,right: 0,bottom: 0} |
