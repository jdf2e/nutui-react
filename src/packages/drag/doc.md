# Drag 拖拽

### 介绍

实现可拖拽的任意元素

### 安装

```javascript
import { Drag } from '@nutui/nutui-react'
```

## 代码演示

### 基础用法

## 基本用法

```tsx
<Drag>
  <div class="touch-dom">触摸移动</div>
</Drag>
```

## 限制拖拽方向

```tsx
<Drag direction={'x'}>
  <div class="touch-dom">只能在X轴拖动</div>
</Drag>
<Drag direction={'y'}>
  <div class="touch-dom">只能在Y轴拖动</div>
</Drag>
```

## 自动吸边

```tsx
<Drag direction={'x'} attract={true}>
  <div class="touch-dom">拖动我</div>
</Drag>
```

## 限制拖拽边界

```tsx
const right = () => {
  return document.documentElement.clientWidth - 300 - 9
}
const bottom = () => {
  return document.documentElement.clientHeight - 559
}
<Drag boundary={{ top: 361, left: 9, bottom: bottom(), right: right() }}>
  <div class="touch-dom">拖动我</div>
</Drag>
```

## API

### Props

| 参数 | 说明 | 类型 | 默认值 |
| --------------- | ----------------------------- | ------- | ------ |
| attract | 是否开启自动吸边 | Boolean | false |
| direction | 拖拽元素的拖拽方向限制，x、y、all三选一 | String |'all' |
| boundary | 拖拽元素的拖拽边界 | Object | {top: 0,left: 0,right: 0,bottom: 0} |
