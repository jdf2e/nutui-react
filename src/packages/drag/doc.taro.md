# Drag 拖拽

## 介绍

实现可拖拽的任意元素

## 安装

```tsx
import { Drag } from '@nutui/nutui-react-taro'
```

## 代码演示

### 基础用法

:::demo

```tsx
import  React from "react";
import { Drag } from '@nutui/nutui-react-taro';

const App = () => {
  return (
    <Drag>
        <div className="touch-dom">触摸移动</div>
    </Drag>
 );
};
export default App;
```

:::

### 限制拖拽方向

:::demo

```tsx
import  React from "react";
import { Drag, Button } from '@nutui/nutui-react-taro';

const App = () => {
  return (
    <>
    <Drag direction="x" style={{ top: '200px', left: '8px' }}>
        <Button type="primary">只能X轴拖拽</Button>
      </Drag>
      <Drag direction="y" style={{ top: '200px', right: '50px' }}>
        <Button type="primary">只能Y轴拖拽</Button>
      </Drag>
    </>
 );
};
export default App;
```

:::

### 自动吸边

:::demo

```tsx
import  React from "react";
import { Drag } from '@nutui/nutui-react-taro';

const App = () => {
  return (
    <Drag direction="x" attract>
      <div className="touch-dom" >按钮</div>
    </Drag>
 );
};
export default App;
```

:::

### 限制拖拽边界

:::demo

```tsx
import  React from "react";
import { Drag, Button } from '@nutui/nutui-react-taro';

const App = () => {
    const right = () => {
    return document.documentElement.clientWidth - 300 - 9
    }
    const bottom = () => {
        return document.documentElement.clientHeight - 202
    }
    return (
    <>
    <div
        className="drag-boundary"
        style={{
          position: 'absolute',
          top: '0px',
          left: '8px',
          width: '300px',
          height: '200px',
          border: '1px solid red',
        }}
       />
      <Drag
        boundary={{ top: 1, left: 9, bottom: bottom(), right: right() }}
        style={{ top: '40px', left: '50px' }}
      >
        <Button type="primary">限制拖拽边界</Button>
      </Drag>
    </>
   );
};
export default App;
```

:::

## Drag

### Props

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| attract | 是否开启自动吸边 | `boolean` | `false` |
| direction | 拖拽元素的拖拽方向限制 | `x` \| `y` \| `all` | `all` |
| boundary | 拖拽元素的拖拽边界 | `Object` | `{top: 0, left: 0, right: 0, bottom: 0}` |