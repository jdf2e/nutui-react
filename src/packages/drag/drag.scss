.nut-drag {
  position: fixed;
  display: inline-flex;
  z-index: 9997 !important;
  width: fit-content;
  height: fit-content;
  touch-action: none;
  user-select: none;
  font-size: 0;
}
