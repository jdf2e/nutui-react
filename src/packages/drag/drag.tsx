import React, { FunctionComponent, useState, useEffect, useRef } from 'react'
import { useDrag } from '@use-gesture/react'
import { useSpring, animated } from '@react-spring/web'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

export interface DragProps extends BasicComponent {
  attract: boolean
  direction: 'x' | 'y' | 'lock' | undefined
  boundary: {
    top: number
    left: number
    right: number
    bottom: number
  }
}
const defaultProps = {
  ...ComponentDefaults,
  attract: false,
  direction: undefined,
  boundary: {
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  },
} as DragProps
export const Drag: FunctionComponent<
  Partial<DragProps> & React.HTMLAttributes<HTMLDivElement>
> = (props) => {
  const { attract, direction, boundary, children, className, style, ...reset } =
    {
      ...defaultProps,
      ...props,
    }
  const classPrefix = 'nut-drag'
  const [boundaryState, setBoundaryState] = useState(boundary)
  const myDrag = useRef<HTMLDivElement>(null)
  const [currstyle, api] = useSpring(() => ({
    x: 0,
    y: 0,
  }))

  const middleLine = useRef(0)

  const getInfo = () => {
    const el = myDrag.current
    if (el) {
      const { offsetWidth, offsetHeight, offsetTop, offsetLeft } = el
      const { clientWidth, clientHeight } = document.documentElement
      const { top, left, bottom, right } = boundary
      setBoundaryState({
        top: -offsetTop + top,
        left: -offsetLeft + left,
        bottom: clientHeight - offsetHeight - offsetTop - bottom,
        right: clientWidth - offsetWidth - offsetLeft - right,
      })
      middleLine.current =
        clientWidth - offsetWidth - offsetLeft - (clientWidth - offsetWidth) / 2
    }
  }

  const bind = useDrag(
    ({ down, last, offset: [x, y] }) => {
      api.start({ x, y, immediate: down })
      if (last) {
        if (direction !== 'y' && attract) {
          if (x < middleLine.current) {
            api.start({ x: boundaryState.left, y, immediate: down })
          } else {
            api.start({
              x: boundaryState.right,
              y,
              immediate: down,
            })
          }
        }
      }
    },
    {
      from: () => [currstyle.x.get(), currstyle.y.get()],
      axis: direction,
      bounds: boundaryState,
    }
  )

  useEffect(() => {
    getInfo()
  }, [myDrag])

  return (
    <div
      style={style}
      className={`${classPrefix} ${className}`}
      {...reset}
      ref={myDrag}
    >
      <animated.div style={currstyle} {...bind()}>
        {children}
      </animated.div>
    </div>
  )
}

Drag.displayName = 'NutDrag'
