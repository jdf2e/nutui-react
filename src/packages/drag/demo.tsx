import React from 'react'
import { Drag } from './drag'
import { useTranslate } from '../../sites/assets/locale'

interface T {
  basic: string
  dragBasic: string
  direction: string
  directionX: string
  directionY: string
  attract: string
  attractText: string
  limitBoundaries: string
}
const DragDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      basic: '基本用法',
      dragBasic: '触摸移动',
      direction: '限制拖拽方向',
      directionX: '只能X轴拖动',
      directionY: '只能Y轴拖动',
      attract: '自动吸边',
      attractText: '按钮',
      limitBoundaries: '限制拖拽边界',
    },
    'zh-TW': {
      basic: '基本用法',
      dragBasic: '觸摸移動',
      direction: '限製拖拽方向',
      directionX: '只能X軸拖動',
      directionY: '只能Y軸拖動',
      attract: '自動吸邊',
      attractText: '按鈕',
      limitBoundaries: '限製拖拽邊界',
    },
    'en-US': {
      basic: 'Basic Usage',
      dragBasic: 'Button',
      direction: 'Limit Direction',
      directionX: 'X axis',
      directionY: 'Y axis',
      attract: 'Attract',
      attractText: 'Button',
      limitBoundaries: 'Limit Boundaries',
    },
  })
  const right = () => {
    return document.documentElement.clientWidth - 300 - 9
  }
  const bottom = () => {
    return document.documentElement.clientHeight - 559
  }
  const btnStyle = {
    borderRadius: '25px',
    padding: '0 18px',
    fontSize: '14px',
    color: '#fff',
    display: 'inline-block',
    lineHeight: '36px',
    background: 'linear-gradient(135deg,#fa2c19 0,#fa6419 100%)',
  }
  return (
    <div className="demo">
      <h2>{translated.basic}</h2>
      <Drag style={{ top: '120px', left: '8px' }}>
        <span style={btnStyle}>{translated.dragBasic}</span>
      </Drag>
      <h2 style={{ top: '30px', position: 'relative' }}>
        {translated.direction}
      </h2>
      <Drag direction="x" style={{ top: '200px', left: '8px' }}>
        <span style={btnStyle}> {translated.directionX}</span>
      </Drag>
      <Drag direction="y" style={{ top: '200px', right: '50px' }}>
        <span style={btnStyle}> {translated.directionY}</span>
      </Drag>
      <h2 style={{ top: '60px', position: 'relative' }}>
        {translated.attract}
      </h2>
      <Drag direction="x" attract style={{ top: '275px', left: '8px' }}>
        <span style={btnStyle}>{translated.attractText}</span>
      </Drag>
      <h2 style={{ top: '90px', position: 'relative' }}>
        {translated.limitBoundaries}
      </h2>
      <div
        className="drag-boundary"
        style={{
          position: 'absolute',
          top: '360px',
          left: '8px',
          width: '300px',
          height: '200px',
          border: '1px solid red',
        }}
      />
      <Drag
        boundary={{ top: 361, left: 9, bottom: bottom(), right: right() }}
        style={{ top: '400px', left: '50px' }}
      >
        <span style={btnStyle}>{translated.limitBoundaries}</span>
      </Drag>
    </div>
  )
}

export default DragDemo
