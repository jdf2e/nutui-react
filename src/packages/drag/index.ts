import { Drag } from './drag'

export type { DragProps } from './drag'
export default Drag
