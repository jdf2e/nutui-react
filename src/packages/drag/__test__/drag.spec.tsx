import * as React from 'react'
import { fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Drag } from '../drag'
import { sleep } from '@/utils/sleep'

describe('Drag', () => {
  const originWindowProto = Object.getPrototypeOf(window)
  beforeAll(() => {
    Object.setPrototypeOf(window, Window.prototype)
  })
  afterAll(() => {
    Object.setPrototypeOf(window, originWindowProto)
  })
  const btnStyle = {
    borderRadius: '25px',
    padding: '0 18px',
    fontSize: '14px',
    color: '#fff',
    display: 'inline-block',
    lineHeight: '36px',
    background: 'linear-gradient(135deg,#fa2c19 0,#fa6419 100%)',
  }

  test('should render default slot correctly', () => {
    const { container } = render(
      <>
        <Drag style={{ top: '120px', left: '8px' }}>
          <span className="custom-inner" style={btnStyle}>
            自定义内容
          </span>
        </Drag>
      </>
    )
    expect(container.querySelector('.custom-inner')?.innerHTML).toBe(
      '自定义内容'
    )
    expect(container.querySelector('.custom-inner')).toMatchSnapshot()
  })

  test('touch move', async () => {
    const { container } = render(
      <>
        <Drag style={{ top: '120px', left: '8px' }}>
          <span className="custom-inner" style={btnStyle}>
            按钮
          </span>
        </Drag>
      </>
    )
    const dragInnerEle = container.querySelector('.nut-drag>div') as Element
    fireEvent.mouseDown(dragInnerEle, {
      buttons: 1,
    })
    fireEvent.mouseMove(dragInnerEle, {
      buttons: 1,
      clientX: 20,
      clientY: 20,
      offset: [100, 100],
    })
    fireEvent.mouseUp(dragInnerEle)

    await sleep(1000)
    expect(dragInnerEle).toMatchSnapshot()
  })
})
