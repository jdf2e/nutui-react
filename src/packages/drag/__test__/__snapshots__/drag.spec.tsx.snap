// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drag should render default slot correctly 1`] = `
<span
  class="custom-inner"
  style="border-radius: 25px; padding: 0px 18px; font-size: 14px; color: rgb(255, 255, 255); display: inline-block; line-height: 36px;"
>
  自定义内容
</span>
`;

exports[`Drag touch move 1`] = `
<div
  style="transform: none;"
>
  <span
    class="custom-inner"
    style="border-radius: 25px; padding: 0px 18px; font-size: 14px; color: rgb(255, 255, 255); display: inline-block; line-height: 36px;"
  >
    按钮
  </span>
</div>
`;
